csharp/src/ListView.cs

```diff
@@ -0,0 +1,178 @@
+// This file is part of the ArmoniK project
+//
+// Copyright (C) ANEO, 2022-2023.All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+using JetBrains.Annotations;
+
+namespace ArmoniK.Utils;
+
+/// <summary>
+///   View over a list of <typeparamref name="Tsrc" />.
+///   All elements are projected into <typeparamref name="Tdst" />.
+/// </summary>
+/// <typeparam name="Tsrc">Type of the underlying list elements</typeparam>
+/// <typeparam name="Tdst">Type of the projected elements</typeparam>
+/// <remarks>
+///   If the underlying list is modified, the view will reflect the changes.
+///   If it is safe to iterate over the underlying list while modifying it,
+///   it is safe to iterate over the view while modifying the underlying list.
+/// </remarks>
+/// <remarks>
+///   The projection should be pure (no side effect).
+/// </remarks>
+[PublicAPI]
+public readonly struct ListView<Tsrc, Tdst> : IList<Tdst>
+{
+  private readonly IList<Tsrc>      list_;
+  private readonly Func<Tsrc, Tdst> projection_;
+
+  /// <summary>
+  ///   Project <paramref name="list" /> into a view
+  /// </summary>
+  /// <param name="list">List to project</param>
+  /// <param name="projection">
+  ///   Function to project <typeparamref name="Tsrc" /> elements into <typeparamref name="Tdst" />
+  ///   elements
+  /// </param>
+  [PublicAPI]
+  public ListView(IList<Tsrc>      list,
+                  Func<Tsrc, Tdst> projection)
+  {
+    projection_ = projection;
+    list_       = list;
+  }
+
+  /// <inheritdoc />
+  public IEnumerator<Tdst> GetEnumerator()
+    => list_.Select(projection_)
+            .GetEnumerator();
+
+  /// <inheritdoc />
+  IEnumerator IEnumerable.GetEnumerator()
+    => GetEnumerator();
+
+  /// <inheritdoc />
+  /// <remarks>
+  ///   <see cref="Add" /> is not supported on a view
+  /// </remarks>
+  public void Add(Tdst item)
+    => throw new NotSupportedException("View List is read-only");
+
+  /// <inheritdoc />
+  /// <remarks>
+  ///   <see cref="Clear" /> is not supported on a view
+  /// </remarks>
+  public void Clear()
```

Comment:
Should you add that a `NotSupportedException` can be thrown by this method in xml-doc or is it already there ?

Reply:
Yes, it is already in the parent documentation: https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.icollection-1.clear?view=net-8.0#exceptions